### PR TITLE
LYN-6431 Added materials support to TerrainPhysicsCollider, added uni…

### PR DESCRIPTION
**Pull request by sphrose**
…t tests.

Previously reviewed in https://github.com/aws-lumberyard-dev/o3de/pull/260, remade PR to enable push to alternative destination.

Signed-off-by: sphrose <82213493+sphrose@users.noreply.github.com>

### CODE REVIEW

File: Gems/Terrain/Code/Tests/TerrainPhysicsColliderTests.cpp [hunk 1]
```diff
@@ -290,3 +290,134 @@ TEST_F(TerrainPhysicsColliderComponentTest, TerrainPhysicsColliderReturnsRelativ
...
+    ON_CALL(terrainListener, GetTerrainHeightQueryResolution).WillByDefault(Return(mockHeightResolution));
+    ON_CALL(terrainListener, GetHeightFromFloats).WillByDefault(Return(mockHeight));
+    ON_CALL(terrainListener, GetMaxSurfaceWeightFromFloats).WillByDefault(
+            [return1, return2](
+                [[maybe_unused]] float x, [[maybe_unused]] float y,
+                [[maybe_unused]] AzFramework::Terrain::TerrainDataRequests::Sampler sampleFilter,
+                [[maybe_unused]] bool* terrainExistsPtr)
+            {
+                // return tag1 for the first half of the rows, tag2 for the rest.
+                if (y < 128)
```
[the remark this answers is not in the document]
Comment (sphrose) [reply Gems/Terrain/Code/Tests/TerrainPhysicsColliderTests.cpp:396]: Fixed

File: Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp [hunk 2]
```diff
@@ -16,16 +16,47 @@
...
+        {
+            serialize->Class<TerrainPhysicsSurfaceMaterialMapping>()
+                ->Version(1)
+                ->Field("Surface", &TerrainPhysicsSurfaceMaterialMapping::m_surfaceTag)
+                ->Field("Material", &TerrainPhysicsSurfaceMaterialMapping::m_materialId);
+
+            if (auto edit = serialize->GetEditContext())
+            {
+                edit->Class<TerrainPhysicsSurfaceMaterialMapping>(
+                        "Terrain Surface Gradient Mapping", "Mapping between a surface and a physics material.")
```
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp:36]: Nit: "Terrain Surface Gradient Mapping" -> "Terrain Surface Material Mapping"

File: Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp [hunk 3]
```diff
@@ -35,7 +66,11 @@ namespace Terrain
                         "Provides terrain data to a physics collider with configurable surface mappings.")
                     ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
                     ->Attribute(AZ::Edit::Attributes::Visibility, AZ::Edit::PropertyVisibility::ShowChildrenOnly)
-                    ->Attribute(AZ::Edit::Attributes::AutoExpand, true);
+                    ->Attribute(AZ::Edit::Attributes::AutoExpand, true)
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default, &TerrainPhysicsColliderConfig::m_surfaceMaterialMappings,
+                        "Surface to Material Mappings ", "Maps surfaces to physics materials")
```
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp:72]: Nit: Extra space after Mappings

File: Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp [hunk 4]
```diff
@@ -248,6 +283,22 @@ namespace Terrain
             }
         }
     }
+    
+    uint8_t TerrainPhysicsColliderComponent::FindSurfaceTagIndex(const SurfaceData::SurfaceTag tag) const
+    {
+        uint8_t index = 0;
+
+        for (auto mapping : m_configuration.m_surfaceMaterialMappings)
```
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp:291]: auto mapping -> should be auto&

File: Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp [hunk 5]
```diff
@@ -286,9 +337,19 @@ namespace Terrain
...
+                    surfaceWeight, &AzFramework::Terrain::TerrainDataRequests::GetMaxSurfaceWeightFromFloats, x, y,
+                    AzFramework::Terrain::TerrainDataRequests::Sampler::DEFAULT, nullptr);
+                    
                 Physics::HeightMaterialPoint point;
                 point.m_height = height - worldCenterZ;
                 point.m_quadMeshType = terrainExists ? Physics::QuadMeshType::SubdivideUpperLeftToBottomRight : Physics::QuadMeshType::Hole;
+                point.m_materialIndex = FindSurfaceTagIndex(surfaceWeight.m_surfaceType);
+                
+                AZ_Warning("TerrainPhysicsCollider", point.m_materialIndex != InvalidSurfaceTagIndex, "SurfaceTag not assigned.");
+
```
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp:352]: No need to warn on this, it's valid to have a surface tag that isn't mapped.  However, to deal with this, it looks like we'll need to do another task that was on our "to-do" list - add a default physics material to the component as well.  FindSurfaceTagIndex() would then return the default material index if there wasn't a specific tag mapping set up.

File: Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp [hunk 6]
```diff
@@ -332,7 +393,15 @@ namespace Terrain
 
     AZStd::vector<Physics::MaterialId> TerrainPhysicsColliderComponent::GetMaterialList() const
     {
-        return AZStd::vector<Physics::MaterialId>();
+        AZStd::vector<Physics::MaterialId> materialList;
+        materialList.reserve(m_configuration.m_surfaceMaterialMappings.size());
+
+        for (auto mapping : m_configuration.m_surfaceMaterialMappings)
```
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp:399]: auto -> auto&

**Review by tjmichaels — APPROVED** [review]

**Pull request merged**